Mathlib/Tactic/Recover.lean

```diff
@@ -0,0 +1,72 @@
+/-
+Copyright (c) 2022 Siddhartha Gadgil. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Gabriel Ebner
+-/
+import Lean
+import Mathlib.Tactic.Cache
+import Mathlib.Tactic.RCases
+import Std
+
+open Std (HashSet PHashSet)
+open Lean Meta Elab Tactic
+
+/- The code below (up to the tactic) is from Aesop due to Janis Limperg-/
+namespace Mathlib.Tactic
+
+namespace Std.HashSet
+
+/-- insert many elements in a HashSet (from Aesop) -/
+def insertMany [ForIn Id ρ α] [BEq α] [Hashable α] (s : HashSet α) (as : ρ) :
+    HashSet α := Id.run do
+  let mut s := s
+  for a in as do
+    s := s.insert a
+  return s
+
+end Std.HashSet
+
+/-- instantiating MVars in goal (from Aesop) -/
+def instantiateMVarsInGoal (mvarId : MVarId) : MetaM Unit := do
+  discard $ getMVarDecl mvarId
+    -- The line above throws an error if the `mvarId` is not declared. The line
+    -- below panics.
+  instantiateMVarDeclMVars mvarId
+
+/-- Returns unassigned metavariable dependencies (from Aesop) -/
+partial def getUnassignedGoalMVarDependencies (mvarId : MVarId) :
```

Comment:
Now that I think about it: does `recover` actually want to recurse into the found metavariables? If we make the 'dropped' metavariables goals, any mvars occurring in them are not dropped any more, so they don't need to also become goals. cc @gebner

Reply:
A situation I can imagine: a new metavariable is created, a dropped mvar assigned to it and removed from goals but the new mvar is not added to goals.

Reply:
If I understand correctly, your situation looks like this:

- initial goal: `?g[?m1]` (this is my notation for '`?g` contains `?m`')
- tactic introduces `?m2` and assigns `?m1 := ?m2`
- tactics returns goals: `[]`

Now if we just collect unassigned mvars in `?g`, we correctly find `?m2` (and not `?m1` since it gets substituted away by `instantiateMVars`). So we set goals:

- goals after `recover`: `[?m2]`

If there's another mvar `?m3` with `?m2[?m3]`, this is fine because `?m3` is already reachable from the tactic state after `recover`. So we don't need to recursively get mvars from `?m2`. However, maybe we want to do it anyway because it's nicer to have an mvar in the goal list than to not have it?

Reply:
I think I misunderstood your point. Is what you are saying that we should collect unassigned mvars in the original goals list, but no need to do this recursively. That sounds reasonable. I will make the change if you confirm that is what you mean and feel it will make things better.

Reply:
Yes exactly, but I'm not sure about this. Let's wait and see what Gabriel thinks.

Reply:
Sure.

Reply:
It's not actually necessary to recurse into the initial goals at all, that was just an optimization. It suffices to just hold on to the list of goals itself and do all recursing through assignments at the end of the block.